src/plugins/ERC20TokenLimitPlugin.sol

```diff
@@ -0,0 +1,170 @@
+/// token contract
+contract ERC20TokenLimitPlugin is BasePlugin, IExecutionHook {
+    using UserOperationLib for PackedUserOperation;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    string public constant NAME = "ERC20 Token Limit Plugin";
+    string public constant VERSION = "1.0.0";
+    string public constant AUTHOR = "ERC-6900 Authors";
+
+    struct ERC20SpendLimit {
+        address token;
+        uint256[] limits;
+    }
+
+    mapping(address account => mapping(address => uint256[] limits)) public limits;
+    mapping(address account => EnumerableSet.AddressSet) internal _tokenList;
+
+    error ExceededNativeTokenLimit();
+    error ExceededNumberOfEntities();
+    error SelectorNotAllowed();
+
+    function getTokensForAccount(address account) external view returns (address[] memory tokens) {
+        tokens = new address[](_tokenList[account].length());
+        for (uint256 i = 0; i < _tokenList[account].length(); i++) {
+            tokens[i] = _tokenList[account].at(i);
+        }
+        return tokens;
+    }
+
+    function updateLimits(uint8 functionId, address token, uint256 newLimit) external {
+        _tokenList[msg.sender].add(token);
+        limits[msg.sender][token][functionId] = newLimit;
+    }
+
+    function _checkAndDecrementLimit(uint8 functionId, bytes4 selector, uint256 spend, address token) internal {
+        if (selector == IERC20.transfer.selector || selector == IERC20.approve.selector) {
+            uint256 limit = limits[msg.sender][token][functionId];
+            if (spend > limit) {
+                revert ExceededNativeTokenLimit();
+            }
+            limits[msg.sender][token][functionId] = limit - spend;
+        } else {
+            revert SelectorNotAllowed();
+        }
+    }
+
+    /// @inheritdoc IExecutionHook
+    function preExecutionHook(uint8 functionId, bytes calldata data) external override returns (bytes memory) {
+        bytes calldata topLevelCallData;
+        bytes4 topLevelSelector;
+
+        // TODO: plugins should never have to do these gymnastics
+        topLevelSelector = bytes4(data[52:56]);
+        if (topLevelSelector == IAccountExecute.executeUserOp.selector) {
+            topLevelCallData = data[56:];
+            topLevelSelector = bytes4(topLevelCallData);
+        } else {
+            topLevelCallData = data[52:];
+        }
+
+        if (topLevelSelector == IStandardExecutor.execute.selector) {
+            address token = address(uint160(uint256(bytes32(topLevelCallData[4:36]))));
+            if (_tokenList[msg.sender].contains(token)) {
+                bytes calldata executeCalldata;
+                uint256 offset = uint256(bytes32(topLevelCallData[68:100]));
+
+                assembly {
+                    let relativeOffset := add(add(topLevelCallData.offset, offset), 4)
+                    executeCalldata.offset := add(relativeOffset, 32)
+                    executeCalldata.length := calldataload(relativeOffset)
```

Comment:
@adam-alchemy: I think this is vulnerable to manipulation because unlike abi.decode, this doesn't check the bounds of the offset or length - you could pass an offset that points to something invalid in the outer calldata, but points to something valid in inner calldata. E.g. the add(...) operations are unchecked, so you could use large numbers to effective subtract an offset, rather than add, and get varying behavior between the checking plugin and the actual receiving token contract.

Might be fine for now with some warnings, or we could just switch these to be abi.decode instead. As a sample, we don't have to worry too much about gas.

Reply:
good catch! will explore abi decode, if it doesn't work will just add a warning since this is for demo purposes, not for prod